### PR TITLE
[3.0] Manage desktop more intelligently.

### PR DESCRIPTION
**Pull request by mtwebster**
- Allow right-click actions on empty desktops
- per-monitor Home folder containers
- new setting to manage primary only, remaining only, or both, or none (show-desktop-icons deprecated)
- new setting to control spill-over of icons from currently non-existent monitors

### CODE REVIEW

File: libnemo-private/nemo-file-operations.c [hunk 1]
```diff
@@ -1076,9 +1077,11 @@ init_common (gsize job_size,
 	common->time = g_timer_new ();
 	common->inhibit_cookie = -1;
 	common->screen_num = 0;
+    common->monitor_num = 0;
 	if (parent_window) {
 		screen = gtk_widget_get_screen (GTK_WIDGET (parent_window));
 		common->screen_num = gdk_screen_get_number (screen);
+        common->monitor_num = nemo_desktop_utils_get_monitor_for_widget (GTK_WIDGET (parent_window));
```
Comment (daschuer) [inline libnemo-private/nemo-file-operations.c]: Missing include: 
#include "nemo-desktop-utils.h"

File: libnemo-private/nemo-global-preferences.h [hunk 2]
```diff
@@ -214,7 +214,9 @@ typedef enum
 #define NEMO_PREFERENCES_LOCKDOWN_COMMAND_LINE         "disable-command-line"
 
 /* Desktop background */
-#define NEMO_PREFERENCES_SHOW_DESKTOP		   "show-desktop-icons"
+#define NEMO_PREFERENCES_DESKTOP_LAYOUT "desktop-layout"
+#define NEMO_PREFERENCES_SHOW_ORPHANED_DESKTOP_ICONS "show-orphaned-desktop-icons"
+#define NEMO_PREFERENCES_SHOW_DESKTOP   "show-desktop-icons"    /* DEPRECATED */
```
Comment (daschuer) [inline libnemo-private/nemo-global-preferences.h]: This feature is currently not only deprecated, but removed. 
This removes also the flexibility of using other file managers than Nemo with Mint cinnamon and it removes also the option to use Nemo in other desktop environments. 

Something similar described here:
http://askubuntu.com/questions/301089/no-desktop-icons-after-uninstalling-nemo
http://www.noobslab.com/2014/02/set-nemo-as-default-file-manager-in.html

How will it be possible to allow these combinations without this "NEMO_PREFERENCES_SHOW_DESKTOP" in future? 

I think it is not a big deal, to not start the new desktop manager conditional by this flag, or do I miss something?

File: src/nemo-application.c [hunk 3]
```diff
@@ -1049,24 +905,7 @@ init_desktop (NemoApplication *self)
 	/* Initialize the desktop link monitor singleton */
 	nemo_desktop_link_monitor_get ();
 
-	if (!self->priv->no_desktop &&
```
Comment (daschuer) [inline src/nemo-application.c]: The no-desktop command line option is now broken.

Comment (daschuer) [inline src/nemo-application.c]: By the way, Nautilus has improved this a bit more and offers "force-desktop" along with "no-desktop"

File: libnemo-private/nemo-global-preferences.h (hunk 2 above)
Comment (mtwebster) [inline libnemo-private/nemo-global-preferences.h]: This is temporary, I plan on restoring this before 3.0 (as well as add some detection to not try to manage the desktop if something else already is) - it's been noted in other reports already.

File: src/nemo-application.c [hunk 4]
```diff
@@ -1049,24 +905,7 @@ init_desktop (NemoApplication *self)
...
-	if (!self->priv->no_desktop) {
-		nemo_application_open_desktop (self);
-	}
-
-	/* Monitor the preference to show or hide the desktop */
-	g_signal_connect_swapped (nemo_desktop_preferences, "changed::" NEMO_PREFERENCES_SHOW_DESKTOP,
-				  G_CALLBACK (desktop_changed_callback),
-				  self);
-
-	g_signal_connect (screen, "monitors-changed",
```
Comment (daschuer) [inline src/nemo-application.c]: screen seams to be unused now.

File: src/nemo-blank-desktop-window.c [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+            gtk_menu_shell_append (GTK_MENU_SHELL (window->details->popup_menu), item);
+        }
+    }
+}
+
+static void
+do_popup_menu (NemoBlankDesktopWindow *window, GdkEvent *event)
+{
+    build_menu (window);
+    eel_pop_up_context_menu (GTK_MENU(window->details->popup_menu),
```
Comment (daschuer) [inline src/nemo-blank-desktop-window.c]: Missing include:
#include <eel/eel-gtk-extensions.h>

File: src/nemo-blank-desktop-window.c [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+            gtk_widget_set_visible (item, TRUE);
+            g_signal_connect (item, "activate", G_CALLBACK (action_activated_callback), action);
+            gtk_menu_shell_append (GTK_MENU_SHELL (window->details->popup_menu), item);
+        }
+    }
+}
+
+static void
+do_popup_menu (NemoBlankDesktopWindow *window, GdkEvent *event)
```
Comment (daschuer) [inline src/nemo-blank-desktop-window.c]: GdkEventButton *

**Comment by norbusan** [comment]
It seems that this commit has created a regression wrt to blinking icons, see https://github.com/linuxmint/nemo/issues/1177 where someone bisected. Is there any change that the original author of the pull request can look into this issue?

**Pull request merged**